### ISSUE TITLE
Modifier-only shortcut

### ISSUE THREAD

**Issue by lwouis** [Kentzo/ShortcutRecorder#114]
I'm building [an app](https://github.com/lwouis/alt-tab-macos/) that triggers when `⌥⇥` is pressed, and should then reacts when `⌥` is released. I would like to let the user decide which shortcut they want to use.

Using `RecorderControl`, the UI refuses to validate the `⌥` as a key. It seems it is viewed only as a modifier, and can't be its own key. However in my current implementation, I'm using the proper `keyCode` and it's working fine, so I don't think there is any technical limitation at play here. Same thing for `Shortcut(keyEquivalent: "⌥")`: it is returning `nil`.

Why can't a shortcut be just a modifier key like option or command? Is there a reason or is it just an oversight?

**Comment by Kentzo** [comment]
AFAIK Cocoa's controls (`NSMenuItem` and `NSButton`) couldn't have such a shortcut and then no one requested it before.

I think that's a reasonable request though.

**Comment by Kentzo** [comment]
No I wonder what should be the correct value of `-[SRShortcut keyCode]` when there is not one.

One option is to change `keyCode`'s type to a larger fixed-width unsigned integer and use it's maximum value to represent lack of the key code.

**Comment by Kentzo** [comment]
Probably a better approach is to change the type to a UInt32 enum and use its maximum as None.

**Comment by Kentzo** [comment]
@lwouis How would you expect the user to end recording of modifier flags only shortcut? Currently the control relies on pressing a non-modifier key to end it.

**Comment by Kentzo** [comment]
Please take a look at the issue-114 branch.

**Comment by lwouis** [comment]
> Please take a look at the issue-114 branch.

Hey @Kentzo, thanks for working on implementing that feature! That's really nice to see!

I took a look at the branch and its latest commit. There are quite a few changes and it's difficult for me to tell if that's correct since i'm unfamiliar with the codebase. On a short-term scale i can spend time manually testing if there is a way for you to share a build of that branch, or maybe i can build it myself easily? I'll check the build instructions later today when i'm not on my phone. I'm unfamiliar with the CI build/tests also, and that would be a big factor in assessing the branch i imagine

**Comment by Kentzo** [comment]
At this time don't worry on validity of the code, just take a look at the changes from an end user perspective: that the feature is indeed works in a way you'd expect it.

**Comment by lwouis** [comment]
Hey @Kentzo I tested the branch today! Thanks for working on this so swiftly by the way, I really appreciate it! I was a bit overwhelmed recently but I finally found time today to test your work! I think it's a good first draft, however I see some issues:

* The recording UX is "modifiers-only" or "modifiers-optional". It seems I can't have a recording UI that can accept "^" and "^a" for example. It would be nice if the user could type either a modifier-only shortcut, or a modifier+key shortcut. You asked earlier "How would you expect the user to end recording of modifier flags only shortcut?". I think it should be: current logic if the user does modifier+key (key ends capture), and new logic if the user presses modifiers then just releases without any key (all modifiers pressed are added to the shortcut, and capture ends when all modifiers keys are up).

* The recording UX is a bit unintuitive for multiple-modifiers-only shortcuts like "^⌥". If you press "⌥" then it ends right away. You need to first press all modifiers you want except one, then click the UI capture element, then press the last modifiers in the combo, then it will record a shortcut with all modifiers. I think my proposal for ending recording above would address that nicely.

* `Shortcut(keyEquivalent: value)` still returns `nil` for modifier-only shortcuts. This means I can use `RecorderControl` with `allowsModifierFlagsOnlyShortcut = true` to show the user a recording UI, but I can't then use it to construct a `Shortcut` object required to listen to that shortcut later on using `GlobalShortcutMonitor` for instance.

**Comment by Kentzo** [comment]
>   I think it should be: current logic if the user does modifier+key (key ends capture), and new logic if the user presses modifiers then just releases without any key (all modifiers pressed are added to the shortcut, and capture ends when all modifiers keys are up)

That is logically sound. However it may be annoying to an end user. I would like to see a real case where such mix is needed and is helpful.

> The recording UX is a bit unintuitive for multiple-modifiers-only shortcuts like "^⌥".

That is true. This issue can be partially addressed by the delegate of the RecorderControl. However, it won't allow implementations that have to accept both "^" and "⌥".

> Shortcut(keyEquivalent: value) still returns nil for modifier-only shortcuts.

That is an oversight. Good catch! Please use the designated initializer for now.

Side question: did you consider something like a pop-up menu instead for your use case?

**Comment by lwouis** [comment]
> That is an oversight. Good catch! Please use the designated initializer for now.

I'm afraid I don't know which initializer you are referring to here. I'm using `Shortcut(keyEquivalent: value)` in the ShortcutRecorder branch of AltTab. Could you clarify what I should use instead? This is the most important part for us actually, as:

> Side question: did you consider something like a pop-up menu instead for your use case?

We are actually using a dropdown menu at the moment:

![image](https://user-images.githubusercontent.com/106195/69503869-7eb11400-0f61-11ea-98ab-8a203cca4c11.png)
![image](https://user-images.githubusercontent.com/106195/69503864-78229c80-0f61-11ea-9c98-9c8aa09256dc.png)

In the future, I would love to use `RecorderControl` instead so that users can decide to have multiple modifiers to trigger AltTab. However, as a first step, we could just keep the dropdown, and instantiate `Shortcut(keyEquivalent: value)` with the modifiers from the dropdown. It doesn't work in master or in branch 114 at the moment though, unfortunately

**Comment by Kentzo** [comment]
> Could you clarify what I should use instead?

`Shortcut(code: .none, modifierFlags: flag, characters: nil, charactersIgnoringModifiers: nil)`

**Comment by lwouis** [comment]
I updated the code to use that constructor. It works everywhere except one use-case:

When using `GlobalShortcutMonitor` to monitor a modifier-only shortcut, for `forKeyEvent: .up`, it doesn't trigger the shortcut. Shortcuts with modifiers+keycode trigger fine, however modifiers-only like only `.option` don't trigger on key `up`.

**Comment by bryanforbes** [comment]
Has there been any more work on this?

**Comment by Kentzo** [comment]
I'm mostly gathering feedback at this moment.

Please explain your use-case and what's missing in the current implementation.

**Comment by bryanforbes** [comment]
I'm making a program that listens for a global hotkey to mute/unmute the default input device. I want to be able to listen to key combinations like `ctrl-option-command` and `fn-command`.

**Comment by Kentzo** [comment]
Would you like a key down or key up event?

**Comment by bryanforbes** [comment]
I need both for push-to-talk

**Comment by Kentzo** [comment]
When there are multiple modifier flags, it's not straightforward when exactly the shortcut should be invoked. Say you have `fn-command` pressed: should it be invoked as soon as any of the keys is lifted or only after all keys?

**Comment by bryanforbes** [comment]
Thanks for the explanation. In my mind, `key down` would be fired when all of the modifiers are pressed and `key up` when any one of them is lifted (since the shortcut is no longer being pressed). When do `key down` and `key up` fire for something like `cmd-a` currently?

**Comment by Kentzo** [comment]
The up and down events are triggered by the non-modifier key, i.e. `a`. Then modifier flags are checked at this very moment. If it's match, then the shortcut gets invoked.

Consider this scenario: you have a shortcut for `ctrl-shift`, `ctrl-cmd-shift` is currently pressed. Now, depending on the order of of key up events the shortcut may or may not be triggered. This ambiguity is confusing and hard to explain to the end user.

I am thinking about this behavior: the shortcut is determined by the trailing key up-only events. E.g.

`fn down` - `cmd down` - `cmd up` - `fn up` = `fn-cmd`
`fn down` - `cmd down` - `cmd up` - `ctrl down` - `fn up` - `ctrl up` = `fn-ctrl`

What do you think?

**Comment by lwouis** [comment]
I like the logic described by @bryanforbes. In @Kentzo example of `ctrl-shift`, `ctrl-cmd-shift`,

- If both are defined shortcuts:
  - `cmd` up -> the  `ctrl-cmd-shift` shortcut is up-triggered
  - `ctrl` or `shift` up -> both ``ctrl-cmd-shift` and `ctrl-shift` shortcuts are up-triggered
- If only `ctrl-shift` is a defined shortcut:
  - `cmd` up -> nothing happens
  - `ctrl` or `shift` up -> `ctrl-shift` shortcuts is up-triggered

@Kentzo described an alternate logic:

> the shortcut is determined by the trailing key up-only events

I think it's more confusing than @bryanforbes proposal, because the user is able to press some modifiers, release them, but they don't trigger their defined shortcut because other keys were pressed in the flow.

In the example (I assume both `fn-cmd` and `fn-ctrl` are defined shortcuts):

- `fn down` - `cmd down` - `cmd up` - `fn up` = `fn-cmd` -> this one is interesting because I could imagine use-cases that would prefer triggering on "all keys up" instead of "first key up". We could have this be a parameter the modifiers-only shortcut to solve the choice.
- `fn down` - `cmd down` - `cmd up` - `ctrl down` - `fn up` - `ctrl up` = `fn-ctrl` -> this one has the issue I mentioned where the user pressed `cmd` but it had no effect which may confuse them. Here I think this behavior is more direct/simple:

`fn down` - `cmd down` - `cmd up` -> [up-trigger `fn-cmd` if "first-key-up" param was selected] - `ctrl down` - `fn up` -> [up-trigger `fn-cmd` if "all-keys-up" param was selected + up-trigger `fn-ctrl` if "first-key-up" param was selected] - `ctrl up` -> [up-trigger `fn-ctrl` if "all-keys-up" param was selected on the shortcut]

So to recap:

- I think some apps may want an up-trigger when the all keys are down (other keys not in the shortcut can be down too) and one key goes up, and other apps may want to trigger only after all keys are up. That can be a param on the modifiers-only shortcut constructor
- I prefer the more direct logic of looking for: all shortcut keys down -> at this point look for shortcut keys up (either one or all) *regardless* of other keys that could be pressed that are not part of this shortcut; they may influence trigger of another shortcut, but not the current one.

**Comment by Kentzo** [comment]
Agreed, the desired behavior depends on the case at hand. Good start would be to write down a comprehensive list of all key up and key down trigger policies.

**Comment by Kentzo** [comment]
Compiled a small table of use cases. Please let me know if I have missed something.

**Comment by Kentzo** [comment]
@lwouis I'm sorry if you mentioned it elsewhere, but for your usecase do you need ShortcutRecorder to register and handle modifier-only shortcut globally or within your app only (e.g. via `LocalShortcutMonitor`)?

**Comment by lwouis** [comment]
@Kentzo No worries! It's actually a bit tricky to answer that question because the current implementation without ShortcutRecorder has every shortcut global, however, in my ShortcutRecorder branch, I've been experimenting with `LocalShortcutMonitor`. I think for modifier-only, I would only need `LocalShortcutMonitor` to work, however I'm a bit worried that it may not work it the user presses and releases the shortcut really fast, in which case, the window will not show, and thus I'm not sure the `LocalShortcutMonitor` will trigger.

I would say, the local shortcut should be enough, but feel free to implement the global one if it's easier. I can work around with any I think :)

**Comment by Kentzo** [comment]
I'm asking because I struggle to find API to handle global modifier-only shortcuts the same way ShortcutRecorder handles normal shortcuts currently. Currently it *blocks* delivery of the event after it gets handled.

The only option I see now is a combination of LocalShortcutMonitor  and `+[NSEvent addGlobalMonitorForEventsMatchingMask:handler:]` which may not work the way you want it.

**Comment by lwouis** [comment]
@Kentzo i'm afraid i couldn't follow the technicalities you explained 😅

**Comment by Kentzo** [comment]
@lwouis I have altered control's behavior when `allowsModifierFlagsOnlyShortcut` is set to `YES`. Please check it out.

**Comment by Kentzo** [comment]
@lwouis Also added support for `keyCode`-less shortcuts onto `LocalShortcutMonitor` and fixed a number of related issues. You can find some sample code in the Documentation playground, at the end of the Sandbox page.

**Comment by lwouis** [comment]
Thank you @Kentzo! I started testing yesterday and was happy to see the recording part working great (love the way you can press another modifier to add/remove it from the list during capture!). I'll test the `LocalShortcutMonitor` part now 👍

**Comment by lwouis** [comment]
@Kentzo I can confirm that it works great with `LocalShortcutMonitor`!

Unfortunately, I think it doesn't enable my use-case. My current keyboard logic has every interaction a **global** shortcut. I thought I may be able to convert that to: activation with `alt-tab` be global, and then keyboard interactions be **local** (i.e. `LocalShortcutMonitor`) while the UI is open and shown to the user. However that's not possible.

The issue is that while the window is shown to the user, it doesn't have keyboard focus, nor does the app itself. While I could activate the app which would give the window keyboard focus, the app should not steal focus from the current app.

I tried a million approaches with `isKeyWindow`, `isMainWindow`, `acceptsFirstResponder`, `worksWhenModal`, `firstResponder`, and more, including private APIs and SPIs. At the end of the day, it is not possible to send key events without the app/window being really focused, which means we would lose the original focus that existed before pressing `alt-tab`, which is not ok.

The only way forward I can imagine is to use `GlobalShortcutMonitor`. Is it possible for you to implement modifier-only shortcuts for `GlobalShortcutMonitor`?

Some cases I could work around, but the blocker use-case is:

User releases `alt` which should close my app's window. Here I can't send the event to `LocalShortcutMonitor` because my app is not activated. If I activate it, then the previous app loses focus, which is not good.

**Comment by Kentzo** [comment]
@lwouis So far I was not able to modify `GlobalShortcutMonitor` to work without a proper keyCode.

However, you can try `NSEvent/addGlobalMonitorForEvents(matching:handler:)` it is similar to `GlobalShortcutMonitor` except that it doesn't _capture_ the event, merely _taps_ them.

**Comment by lwouis** [comment]
@Kentzo I just tried and it works indeed. I think this may unblock me to integrate for now. However, as you pointed out, it is still not ideal since after the user releases the modifiers-only shortcut, that event is acted upon, but then also forwarded to the previously-focused app.

In my specific case, I don't think many app will react to a modifier-only shortcut being up, so I think it's a reasonable trade-off to adopt ShortcutRecorder, but ideally there would be a way to simply use `GlobalShortcutMonitor.shared` instead.

**Comment by Kentzo** [comment]
Have you also tried to override `NSApplication/sendEvent(_:)`?

**Comment by lwouis** [comment]
Yes I tried. However the NSApplication will only receive events if it is activated in the first place. That activation is what I'm trying to avoid because it loses focus on the original app the user was using

**Comment by Kentzo** [comment]
In that case I doubt there is a better workaround than NSEvent's global monitor. The ancient Carbon API I use to register global hot keys seem ignored to modifier-only shortcuts.

**Comment by lwouis** [comment]
I [currently use](https://github.com/lwouis/alt-tab-macos/blob/master/src/logic/Keyboard.swift#L14) the `CGEvent.tapCreate` API. It's a global monitor that works with any input. It has [some limitations](https://github.com/lwouis/alt-tab-macos/issues?q=is%3Aopen+is%3Aissue+milestone%3Av4) but overall has been serving me great so far. The main goal to move to ShortcutRecorder is the amazing user recording UI elements. I was also thinking that the carbon API used internally was overall superior to `CGEvent.tapCreate`, but this limitation you mentioned made me ask myself: why did you go with the carbon API in the first place? Is there some benefits compared to `CGEvent.tapCreate`? Just curious

**Comment by Kentzo** [comment]
It was inherited from the original implementation of ShortcutRecorder.

**Comment by Kentzo** [comment]
From what I see it appears to be the same as `NSEvent/addGlobalMonitorForEvents(matching:handler:)`, i.e. does not allow to block events.

**Comment by lwouis** [comment]
> From what I see it appears to be the same as `NSEvent/addGlobalMonitorForEvents(matching:handler:)`, i.e. does not allow to block events.

I'm confused. Using ShortcutRecorder `GlobalShortcutMonitor.shared` does block events. It is it using the Carbon API internally, then the Carbon API can block events. I can see in my tests that the keyboard pressed are not propagated to the app with keyboard focus when I used `GlobalShortcutMonitor.shared`

**Comment by Kentzo** [comment]
That is correct, Carbon API does allow to block events.

In my comment regarding `NSEvent/addGlobalMonitorForEvents(matching:handler:)` I compared it against `CGEventTapCreate`.

**Comment by Kentzo** [comment]
Could you explain the limitation of `CGEventTapCreate`? Does or does not it allow blocking the events?

**Comment by lwouis** [comment]
Oh I see, I read it the other way around. `CGEventTapCreate` allows to block the event, or propagate it.

**Comment by lwouis** [comment]
I just found out another limitation of the `addGlobalMonitorForEvents` workaround: it doesn't not handle properly this modifier-only keyUp scenario:

`alt-down` > `tab-down` > UI appears > `tab-up` > `shift-down` > `alt-up` > UI doesn't disappear > `shift-up`

The direct scenario works though:

`alt-down` > `tab-down` > UI appears > `tab-up` > `alt-up` > UI disappears 👍

I imagine, this is because when `addGlobalMonitorForEvents` passes the event to SR, it will be `alt-up` + `shit is down` which is not an exact match, so no trigger.

**Comment by Kentzo** [comment]
Can you share the code of how you configured `LocalShortcutMonitor` and installed NSEvent's monitor?

**Comment by lwouis** [comment]
[This branch](https://github.com/lwouis/alt-tab-macos/blob/shortcut-recorder-wip) builds, and has ShortcutRecorder integrated and working (mostly). Here you can see how I used [LocalShortcutMonitor](https://github.com/lwouis/alt-tab-macos/blob/shortcut-recorder-wip/src/ui/preferences-window/tabs/GeneralTab.swift#L46) and [NSEvent monitors](https://github.com/lwouis/alt-tab-macos/blob/shortcut-recorder-wip/src/ui/preferences-window/tabs/GeneralTab.swift#L88).

**Comment by lwouis** [comment]
I thought that [returning `false` in the ShortcutAction handler](https://github.com/lwouis/alt-tab-macos/blob/shortcut-recorder-wip/src/ui/preferences-window/tabs/GeneralTab.swift#L81) would propagate the event, but it doesn't. For ShortcutActions used with `GlobalShortcutMonitor`, is there a way to not block, but propagate them back to the original application?

I'm using `GlobalShortcutMonitor.shared` so that I can block some shortcuts, unlike `addGlobalMonitorForEvents`. However, sometimes I need to propagate, not always block. Is that possible?

**Comment by Kentzo** [comment]
I'm not sure: is the issue that NSEvent's handler does not receive the necessary flagsChanged event or it's `LocalShortcutMonitor` that fails to dispatch it.

Can you add a breakpoint onto NSEvent's handler with something like `po $0` to record all incoming events and paste the result here?

**Comment by lwouis** [comment]
Here is the code:

```swift
NSEvent.addGlobalMonitorForEvents(matching: .flagsChanged) {
    let x = App.shortcutMonitor.handle($0, withTarget: nil)
    debugPrint($0, x)
}
```

Here is the output when I had `alt-down` and I pressed: `shift-down` > `alt-up` > `shift-up`:

```
NSEvent: type=FlagsChanged loc=(1116.57,543.105) time=264258.0 flags=0x60103 win=0x0 winNum=0 ctxt=0x0 keyCode=56 false
NSEvent: type=FlagsChanged loc=(1116.57,543.105) time=264260.1 flags=0x20102 win=0x0 winNum=0 ctxt=0x0 keyCode=59 false
NSEvent: type=FlagsChanged loc=(1116.57,543.105) time=264261.9 flags=0x100 win=0x0 winNum=0 ctxt=0x0 keyCode=56 false
```

The 3 events correspond to the 3 key events, so `addGlobalMonitorForEvents` is correctly triggering, but `LocalShortcutMonitor.handle` is returning `false` on the second event, but it should not

**Comment by Kentzo** [comment]
What is the value of the `holdShortcut` key equivalent in your code?

**Comment by lwouis** [comment]
> What is the value of the `Preferences.holdShortcut` key equivalent in your code?

It's `⌥`

**Comment by Kentzo** [comment]
In that case, the second flagsChanged event that corresponds to `shift-down & alt-up` is interpreted by `LocalShortcutMonitor` as the `shift-alt` shortcut of the `up` event type.

As you may have noticed, a flagsChanged event has `modifierFlags` equal to the currently pressed modifier keys and `keyCode` equal to whichever modifier key was up or down.

`LocalShortcutMonitor` then [determines](https://github.com/Kentzo/ShortcutRecorder/blob/294dfa7b2664280f795336873cdb28108ed879e5/Library/SRShortcutAction.m#L1248-L1264) whether it was the `down` or `up` event by checking presence of `keyCode` in `modifierFlags`.

**Comment by lwouis** [comment]
This issue is unlikely to really bother people in the real world since it's quite an involved sequence of keys, during which even the user may get confused by what they wanted to do. They would then just do it again and it would be fixed.

However, this is a real problem with large consequences:

> I thought that [returning `false` in the ShortcutAction handler](https://github.com/lwouis/alt-tab-macos/blob/shortcut-recorder-wip/src/ui/preferences-window/tabs/GeneralTab.swift#L81) would propagate the event, but it doesn't. For ShortcutActions used with `GlobalShortcutMonitor`, is there a way to not block, but propagate them back to the original application?
> 
> I'm using `GlobalShortcutMonitor.shared` so that I can block some shortcuts, unlike `addGlobalMonitorForEvents`. However, sometimes I need to propagate, not always block. Is that possible?

I looked everywhere in the docs but couldn't find a way to let the event propagate from the `GlobalShortcutMonitor` handler.

**Comment by Kentzo** [comment]
Pushed a fix onto issue-114 that should make `GlobalShortcutMonitor` to respect result of action's handler.

**Comment by Kentzo** [comment]
Remember, that you can dynamically enable/disable actions even after they are added to a monitor via the `enabled` property.

**Comment by Kentzo** [comment]
@lwouis Looked at your [current solution](https://github.com/lwouis/alt-tab-macos/blob/master/src/logic/Keyboard.swift#L14) for event tapping. I believe you can keep it, just switch to `LocalShortcutMonitor`.

After reading some docs I believe the major difference is that `CGEventTagCreate` requires the process to be either root or have Accessibility enabled. `GlobalShortcutMonitor` does not require that.

**Comment by Kentzo** [comment]
Added the `SRAXGlobalShortcutMonitor` that uses `CGEventTapCreate` under the hood. Let me know if it works for you.

**Comment by lwouis** [comment]
> Added the `SRAXGlobalShortcutMonitor` that uses `CGEventTapCreate` under the hood. Let me know if it works for you.

That's really cool! Nice to have a choice like this for users! I tried to read your implementation, but it's a bit hard for me. I noticed though that there is no mention of "timeout". I remember struggling with random issues until I understood that sometimes the system sends you these 2 events:
https://github.com/lwouis/alt-tab-macos/blob/master/src/logic/Keyboard.swift#L70

You may want to be careful about this, or at least document it to users cause this stops the shortcuts from working if you don't re-enable the tap again

> Pushed a fix onto issue-114 that should make GlobalShortcutMonitor to respect result of action's handler.

I tried returning `false` in the `actionHandler`, and still it blocks the shortcut for the currently focused app. I debugged, and see that [on this line](https://github.com/Kentzo/ShortcutRecorder/commit/7f8fb9fd81bb5f3c223233f787f460d9399f67a2#diff-20128d3661be124be09875a07f104ef6R910), it returns `eventNotHandledErr`, which the docs say is what you should return to let propagate. However like I said it doesn't work in the end.

**Comment by lwouis** [comment]
I think I found a bug in the `issue-114` branch. You changed code around here:

https://github.com/Kentzo/ShortcutRecorder/blob/issue-114/Library/SRShortcut.m#L44

I'm getting exceptions now with `EXC_BAD_INSTRUCTION` because it tries to do `aKeyboardEvent.characters`. Looking at the docs, it says:

> This property is only valid for key-up and key-down events. It raises an NSInternalInconsistencyException if accessed on any other kind of event object.

I don't understand all the logic at play here, but it consistently crashes there after a few minutes when I must be hitting a specific key combination that should not enter that block of code.

**Comment by Kentzo** [comment]
Do you happen to have the full exception with exact trace and message?

**Comment by lwouis** [comment]
Pressing `e` on my keyboard, the event is passed to the monitor `.handle` function which throws:

```
Exception = EXC_BAD_INSTRUCTION (code=EXC_I386_INVOP, subcode=0x0)
self = {Class | 0x1002ead28} SRShortcut
aKeyboardEvent = {NSEvent * | 0x10052a870} 0x000000010052a870
eventType = {NSEventType} NSEventTypeKeyDown
keyCode = {unsigned short} 14
modifierFlags = {NSEventModifierFlags} 256
```

Here is the stack trace:

```
_dispatch_assert_queue_fail 0x00007fff67ff1799
dispatch_assert_queue 0x00007fff67ff172e
islGetInputSourceListWithAdditions 0x00007fff2f4fb952
isValidateInputSourceRef 0x00007fff2f4fe134
TSMGetInputSourceProperty 0x00007fff2f4fdffc
GetCurrentInputSourceKeyboardProperties 0x00007fff2f51e8f3
GetDeadKeyState 0x00007fff2f52c2f6
_convertEventRefToString 0x00007fff2ddb41a5
-[NSEvent characters] 0x00007fff2ddb6c1f
+[SRShortcut shortcutWithEvent:] SRShortcut.m:68
-[SRLocalShortcutMonitor handleEvent:withTarget:] SRShortcutAction.m:1369
AltTab.keyboardHandler(proxy: Swift.OpaquePointer, type: __C.CGEventType, cgEvent: __C.CGEventRef, userInfo: Swift.Optional<Swift.UnsafeMutableRawPointer>) -> Swift.Optional<Swift.Unmanaged<__C.CGEventRef>> Keyboard.swift:34
@objc AltTab.keyboardHandler(proxy: Swift.OpaquePointer, type: __C.CGEventType, cgEvent: __C.CGEventRef, userInfo: Swift.Optional<Swift.UnsafeMutableRawPointer>) -> Swift.Optional<Swift.Unmanaged<__C.CGEventRef>> 0x000000010007701b
processEventTapData(void*, unsigned int, unsigned int, unsigned int, unsigned char*, unsigned int) 0x00007fff5edacc10
_XPostEventTapData 0x00007fff5ef0fc85
eventTapMessageHandler(__CFMachPort*, void*, long, void*) 0x00007fff5edac93b
__CFMachPortPerform 0x00007fff309d4a55
__CFRUNLOOP_IS_CALLING_OUT_TO_A_SOURCE1_PERFORM_FUNCTION__ 0x00007fff309a60bf
__CFRunLoopDoSource1 0x00007fff309a600f
__CFRunLoopRun 0x00007fff309a4c7c
CFRunLoopRunSpecific 0x00007fff309a3bd3
CFRunLoopRun 0x00007fff30a2c19a
closure #1 () -> () in AltTab.listenToGlobalKeyboardEvents() -> () Keyboard.swift:26
reabstraction thunk helper from @escaping @callee_guaranteed () -> () to @escaping @callee_unowned @convention(block) () -> () 0x000000010000509d
_dispatch_call_block_and_release 0x00007fff67fef583
_dispatch_client_callout 0x00007fff67ff050e
_dispatch_lane_serial_drain 0x00007fff67ff5ace
_dispatch_lane_invoke 0x00007fff67ff6452
_dispatch_workloop_worker_thread 0x00007fff67fffa9e
_pthread_wqthread 0x00007fff6824a6fc
start_wqthread 0x00007fff68249827
```

In a more colorful way:

![image](https://user-images.githubusercontent.com/106195/77733565-6f92bd80-704a-11ea-9179-e3b5160bd2f2.png)

**Comment by Kentzo** [comment]
Can you `po aKeyboardEvent`?

**Comment by Kentzo** [comment]
I have been further experimenting with Carbon Hot Key API to find a way to propagate the event when action handlers return false.

I believe it does propagate, however not in a way you expect: the event that gets propagated is of type `kEventHotKeyPressed` / `kEventHotKeyReleased`, not `kEventRawKeyDown` / `kEventRawKeyUp`. In other words the event gets propagated to other hot key handler in the chain.

One workaround is to modify `GlobalShortcutMonitor` to create and re-send an event of appropriate type. However, I believe it requires permissions similar to those of `CGEventTap` which contradicts its purpose of working without requiring anything extra.

**Comment by Kentzo** [comment]
@lwouis Sent [bug report](https://gist.github.com/Kentzo/f8289d99c5d93d81f5ca60b7adc94fae) regarding hot key propagation to Apple. You may want to dupe it.

**Comment by lwouis** [comment]
> Can you run `po aKeyboardEvent`? My guess is that it some other type of event.

I'm afraid it doesn't help much more than the info we already have. Here is what I get from running `po` or clicking the `print info` button in XCode:

```
(lldb) po aKeyboardEvent;
0x0000000107226f80
```

![image](https://user-images.githubusercontent.com/106195/77865363-ee654180-7268-11ea-8707-a356a39d2609.png)

**Comment by Kentzo** [comment]
Hmm, that output from the debugger suggests the event is no longer valid and probably got deallocated in the process. Can you enable zombie objects and see if it gets triggered? Also, please run `po` right after your code receives the event from the system.

**Comment by lwouis** [comment]
@Kentzo I enabled zombie objects, but I don't think it's triggering. I'm not sure what it does actually, and I see nothing new so I assume it doesn't trigger. I put a breakpoint in my app code as soon as I get the `CgEvent` from the `CGEvent.createTap`. Here is what I see:

![image](https://user-images.githubusercontent.com/106195/77887869-4b331d00-72a6-11ea-9595-8afa5b23c9e6.png)

**Comment by Kentzo** [comment]
Can you dump info about `cgEvent` via the `CGEventGet*` family of functions and attach the output? I'll also try to isolate the issue on my machine.

**Comment by Kentzo** [comment]
I was unable to reproduce the exception with the following code:

```
@implementation AppDelegate
{
    CFMachPortRef tap;
    CFRunLoopSourceRef source;
}

CGEventRef _Nullable _SRQuartzEventHandler(CGEventTapProxy aProxy, CGEventType aType, CGEventRef anEvent, void * _Nullable aUserInfo)
{
    if (aType == kCGEventKeyDown || aType == kCGEventKeyUp || aType == kCGEventFlagsChanged)
    {
        NSEvent *e = [NSEvent eventWithCGEvent:anEvent];
        NSLog(@"%@", e);

        if (e.type != NSEventTypeFlagsChanged)
            NSLog(@"%@", e.characters);
    }

    return anEvent;
}

- (void)applicationDidFinishLaunching:(NSNotification *)aNotification {
    static const CGEventMask Mask = (CGEventMaskBit(kCGEventKeyDown) |
                                     CGEventMaskBit(kCGEventKeyUp) |
                                     CGEventMaskBit(kCGEventFlagsChanged));
    NSDictionary *options = @{(__bridge id)kAXTrustedCheckOptionPrompt: @YES};
    BOOL accessibilityEnabled = AXIsProcessTrustedWithOptions((CFDictionaryRef)options);
    NSParameterAssert(accessibilityEnabled);
    tap = CGEventTapCreate(kCGSessionEventTap,
                     kCGHeadInsertEventTap,
                     kCGEventTapOptionDefault,
                     Mask,
                     _SRQuartzEventHandler,
                     (__bridge void *)self);
    source = CFMachPortCreateRunLoopSource(kCFAllocatorDefault, tap, 0);
    CFRunLoopAddSource(NSRunLoop.currentRunLoop.getCFRunLoop, source, kCFRunLoopDefaultMode);
    NSParameterAssert(tap != NULL);
}


- (void)applicationWillTerminate:(NSNotification *)aNotification {
    // Insert code here to tear down your application
}

@end
```

**Comment by lwouis** [comment]
> Can you dump info about `cgEvent` via the `CGEventGet*` family of functions and attach the output? I'll also try to isolate the issue on my machine.

I had to use the ported APIs since the `CGEventGet*` were deprecated and are unusable starting from Swift 3:

![image](https://user-images.githubusercontent.com/106195/77981959-255d5500-7346-11ea-9698-67f3265615af.png)

This is what I got with the ported APIs:

![image](https://user-images.githubusercontent.com/106195/77982500-7457ba00-7347-11ea-8ecf-203578056431.png)

It matches what we saw previously. I think to move forwards I need to make a proof of concept project like you did and see if the issue happens there. I'm also really curious if this issue was happening on previous versions of AltTab, but was simply never an issue because we never use `.characters` on the `NSEvent`. I'll check that

**Comment by lwouis** [comment]
I can reproduce with a single-file app:

```swift
import Cocoa
import Darwin

// OS events should be observed on background threads
class DispatchQueues {
    static let focusActions = DispatchQueue(label: "focusActions", qos: .userInteractive)
    static let keyboardEvents = DispatchQueue(label: "keyboardEvents", qos: .userInteractive)
}

class App: NSApplication, NSApplicationDelegate {
    override init() {
        super.init()
        delegate = self
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }

    var eventTap: CFMachPort?

    func applicationDidFinishLaunching(_ aNotification: Notification) {
        DispatchQueues.keyboardEvents.async {
            let eventMask = [CGEventType.keyDown, CGEventType.keyUp, CGEventType.flagsChanged].reduce(CGEventMask(0), { $0 | (1 << $1.rawValue) })
            // CGEvent.tapCreate returns null if ensureAccessibilityCheckboxIsChecked() didn't pass
            self.eventTap = CGEvent.tapCreate(
                    tap: .cgSessionEventTap,
                    place: .headInsertEventTap,
                    options: .defaultTap,
                    eventsOfInterest: eventMask,
                    callback: keyboardHandler,
                    userInfo: nil)
            let runLoopSource = CFMachPortCreateRunLoopSource(kCFAllocatorDefault, self.eventTap, 0)
            CFRunLoopAddSource(CFRunLoopGetCurrent(), runLoopSource, .defaultMode)
            CFRunLoopRun()
        }
    }
}

func keyboardHandler(proxy: CGEventTapProxy, type: CGEventType, event_: CGEvent, appPointer: UnsafeMutableRawPointer?) -> Unmanaged<CGEvent>? {
    if type == .keyDown || type == .keyUp || type == .flagsChanged {
        if let event = NSEvent(cgEvent: event_) {
            debugPrint("is .keyDown", event.type == .keyDown)
            debugPrint("crashes?", event.characters)
        }
    }
    // focused app will receive the event
    return Unmanaged.passRetained(event_)
}
```

It crashes on the `.characters` call. I'm thinking the issue may be that the RunLoop in on a background thread. Is there any constraint for the CGEvent to be listened to on the Main Thread?

**Comment by lwouis** [comment]
How do you handle multi-threading in ShortcutRecorder? The reason I have a background thread in AltTab is that I want keyboard events to be processes outside the Main Thread. I want them to be able to switch a mutex/flag to stop work on the Main Thread. The use-case is the following:

User hits alt-tab > start rendering the UI to show them > they release alt-tab very quickly > event is received instantly on the background thread while the main thread is busy with rendering the UI > background thread sets a flag to `false` > Main thread regularly checks that flag to decide if it should continue or not > flag is `false` so task on main thread stops and no UI is shown to the user, and instead AltTab switches to the other app instantly.

The purpose is to avoid flickering the UI in the scenario where the user knows they want the previous app, and don't need to see the UI. They will just quickly hit the shortcut, and we want to abort rendering the UI, so it isn't shown for a few frames, making it flicker for the user.

**Comment by lwouis** [comment]
> @lwouis Sent [bug report](https://gist.github.com/Kentzo/f8289d99c5d93d81f5ca60b7adc94fae) regarding hot key propagation to Apple. You may want to dupe it.

There is [interesting info on those WWDC 2019 slides](file:///Users/louis.pontoise/Downloads/701_advances_in_macos_security.pdf). Apple clearly wants to limit what can be done out-of-the-box for macOS apps, and have them ask for permissions to do things such as listening to the keyboard, or sending synthetic events like imitating a click to avoid malware clicking dialog for instance.

In that context, I don't understand how the Carbon API was not banned. Makes 0 sense to me that I can write a keylogger using the Carbon API, but doing the same prompts Keyboard permission with the CGEvent tap API.

It is not directly linked to the bug issue you linked, as in that case I believe you should be able to let the *original* event propagate after interception. I don't see how privacy is a concern here (it is for the whole Carbon API, I would argue), and it seems like a bug.

**Comment by lwouis** [comment]
A bit unrelated but I just found this out: the `allowsEmptyModifierFlags` is not respected. I don't know if it's a regression in the `issue-114` branch, or if it's on `master`, but I have a `RecorderControl` on which I `.set(allowedModifierFlags: CocoaModifierFlagsMask, requiredModifierFlags: [], allowsEmptyModifierFlags: false)`. I can see that the value is correctly set to false: `debugPrint(input.allowsEmptyModifierFlags)` -> `false`. Yet in the UI, I can set the control to just `e` for instance:

![image](https://user-images.githubusercontent.com/106195/77998038-6ae04900-736b-11ea-8093-2aa5db195451.png)
![image](https://user-images.githubusercontent.com/106195/77998054-6e73d000-736b-11ea-8193-9e5400c79608.png)
![image](https://user-images.githubusercontent.com/106195/77998072-73388400-736b-11ea-80fe-c535b8545e64.png)

**Comment by Kentzo** [comment]
> In that context, I don't understand how the Carbon API was not banned. 

I guess the main difference is that Carbon API I use is akin to addGlobalEventMonitor, i.e. allows to receive events only, not alter them. While CGEventTap implies that the event being passed can be modified by the handler.

My understanding that when you register a key combination as a Hot Key (in Carbon terms), the OS assigns a different event type before dispatching it further. Hence the apps never see it as a Key Down / Key Up event.

The bug report I have filed requests to make the OS to retry the event as Key Down / Key Up *if* all Hot Key handlers returned false.

---

> The use-case is the following:

If it is the rendering that is too slow, perhaps you can offload it to the background thread instead? Then you can handler events in the Main Thread and cancel/pause/resume rendering on demand.

> How do you handle multi-threading in ShortcutRecorder?

I do not. I trust the system to deliver the event and call the handler in the most appropriate queue and the user to perform the work in the most appropriate queue.

---

Since you do not need the characters, there is a workaround: instead of using `NSEvent(cgEvent:)` use `NSEvent.keyEvent(…)`; then use `CGEvent.*` properties to get the values for most of the arguments and specify an empty string for `characters` and `charactersIgnoringModifiers`.

**Comment by Kentzo** [comment]
> A bit unrelated but I just found this out: the allowsEmptyModifierFlags is not respected.

What is the value of `CocoaModifierFlagsMask`?

**Comment by lwouis** [comment]
> If it is the rendering that is too slow, perhaps you can offload it to the background thread instead? Then you can handler events in the Main Thread and cancel/pause/resume rendering on demand.

The rendering happens within `NSView` components, which are all not thread-safe and must all be rendered/updated on the main thread. This is a big design constraint coming from Apple's AppKit. The only way to escape that is to do completely custom rendering, which is out of scope for me as I don't want to turn this project into an OpenGL video game engine. I don't have the skillset for that 😅

> I do not. I trust the system to deliver the event and call the handler in the most appropriate queue and the user to perform the work in the most appropriate queue.

But you have to have an event loop somewhere. In my current approach for instance with `CGEvent` taps, I have a `RunLoop` I can place on some thread of my chosing. When I use `GlobalShortcutMonitor` I can decide where the callback gets executed, but there must be an event loop somewhere for `GlobalShortcutMonitor` to receive the events from the OS. Are you saying it happens on the main thread without a way to chance that?

> there is a workaround: instead of using NSEvent(cgEvent:) use NSEvent.keyEvent(…)

That works! Wow what a life-saver!

> What is the value of `CocoaModifierFlagsMask?`

It's coming from SR. Basically all legit modifier keys:

https://github.com/Kentzo/ShortcutRecorder/blob/8e2b5a5792ffc8c6c75834e74ddca7ebf6c460a5/Library/SRCommon.h#L16

**Comment by lwouis** [comment]
> A bit unrelated but I just found this out: the allowsEmptyModifierFlags is not respected.

I think I will implement custom validation to work around that issue. I actually need all non-modifier keys to be not allowed. This means even if `allowsEmptyModifierFlags` was working, I would still need to prevent `modifier(s) + key`. To illustrate:

* `⌥` -> valid
* `e` -> invalid
* `⌥e` -> invalid

I believe there is documentation for custom validators in the project. I'll check it in the playground examples later

**Comment by Kentzo** [comment]
> But you have to have an event loop somewhere.

For `GlobalShortcutMonitor` I rely on API that does not require me to specify a run loop. I assume that Carbon API does not require one or implicitly use a default one from the main thread.

For `AXGlobalShortcutMonitor` I default to the current run loop where the object is instantiated unless you pass custom run loop.

For `LocalShortcutMonitor` there is no need for a run loop as it merely dispatches events and depends on the caller.

**Comment by Kentzo** [comment]
Added a remedy for the NSEvent bug we just discovered.

**Comment by lwouis** [comment]
> I think I will implement custom validation to work around that issue. I actually need all non-modifier keys to be not allowed.

That was very simple to do by overriding the following method in a subclass of `RecorderControl`:

```swift
override func areModifierFlagsValid(_ aModifierFlags: NSEvent.ModifierFlags, for aKeyCode: KeyCode) -> Bool {
        if allowsModifierFlagsOnlyShortcut && aKeyCode != .none {
            return false
        }
        return super.areModifierFlagsValid(aModifierFlags, for: aKeyCode)
}
```

**Comment by lwouis** [comment]
> Added a remedy for the NSEvent bug we just discovered.

I checked the commit, and I think it could be improved. You are trying to access `.characters`, and [catching exceptions if they arise](https://github.com/Kentzo/ShortcutRecorder/commit/646a59c8f53a336456be05fffd5cde4ca1d4328d#diff-b9e87da66f98492095a54f12807902a9R75). In case like mine where the events are processed in a background thread, every single call is going to throw, get caught, then proceed. I would argue that exception should be reserved for un-foreseen events, and here we know about the use-case already, thus you could first do the `NSThread.isMainThread` check, and depending on the context, fallback to empty strings.

I think that would lead to:

* Better performance by not using exceptions for a code hot-path
* Clarify that this is a workaround for a supported use-case, not an unexpected situation

**Comment by Kentzo** [comment]
Exceptions are ["zero-cost"](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/Exceptions/Articles/Exceptions64Bit.html) in objc modern runtime (read 64bit, 10.6+).

The reason I opted for exception is that I believe it's a bug. It may have worked before it may get fixed in the future. ShortcutRecorder does not make any promises regarding thread safety and expects external code to take care of the issue. I suggest you to keep the workaround.

Btw, have you sent the bug report to apple?

**Comment by lwouis** [comment]
To clarify the state of things regarding AltTab:

* I have a working integration of SR that uses the previous `CGEvent` `createTap` code from AltTab together with a `LocalShortcutRecorder`
* Ideally I would like to use `GlobalShortcutMonitor` as I was hoping it would fix these issues: https://github.com/lwouis/alt-tab-macos/issues/17, https://github.com/lwouis/alt-tab-macos/issues/175, https://github.com/lwouis/alt-tab-macos/issues/157
* The shortcut recording is working, with me being able to have controls which take modifiers only, and controls which take modifier(s)+key or just key. This was possible by overriding `areModifierFlagsValid`
* I kept the `CGEvent` code running on a background thread. This was possible with the  `NSEvent.keyEvent` workaround
* The UI looks a bit awkward as the controls are very large:
  ![image](https://user-images.githubusercontent.com/106195/78105941-4f3d7700-742d-11ea-8163-f6ccc0d23e21.png)
  I tried to make them less wide by adjusting the `widthAnchor`:
  ![image](https://user-images.githubusercontent.com/106195/78105960-5bc1cf80-742d-11ea-9cf9-3981af733338.png)
 It looks better, but during interaction, it doesn't show all the keys properly:
  ![Screen Shot 2020-04-01 at 15 29 17](https://user-images.githubusercontent.com/106195/78106092-a17e9800-742d-11ea-8dfe-e649e14da098.jpg)
  Then when I release, I see the actual, full shortcut:
  ![Screen Shot 2020-04-01 at 15 29 36](https://user-images.githubusercontent.com/106195/78106094-a2afc500-742d-11ea-9cac-bb5f3ab2a743.jpg)
Is there a way to have the controls have a smaller minimum width, and then have them grow wider as they need to host more text?
* I'm using the following Podfile code: `pod 'ShortcutRecorder', :git => 'https://github.com/Kentzo/ShortcutRecorder.git', :branch => 'issue-114'`. Should I continue using that, or are you planning a release later on?

**Comment by Kentzo** [comment]
I'm going to merge this branch into master as soon as all issues related to modifier-only shortcuts are tucked.

Have you considered the `enabled` property of shortcuts and ability to `pause`/`resume` the global monitor? Perhaps it can help in implementing your scenario.

**Comment by Kentzo** [comment]
In the screenshots above, what is the exact value of control's width?

**Comment by lwouis** [comment]
> Btw, have you sent the bug report to apple?

I did. I also opened a ticket for the `NSEvent.characters*` access crashing on a background thread.

> In the screenshots above, what is the exact value of control's width?

In the `RecorderControl`:

```swift
debugPrint(frame.width, fittingSize.width, intrinsicContentSize.width)
widthAnchor.constraint(equalToConstant: 80).isActive = true
debugPrint(frame.width, fittingSize.width, intrinsicContentSize.width)
```

Logs:

```
0.0 146.0 144.0
0.0 82.0 144.0
0.0 146.0 144.0
0.0 82.0 144.0
0.0 146.0 144.0
0.0 82.0 144.0
0.0 146.0 144.0
0.0 82.0 144.0
```

Thus I believe the `fittingSize.width` is 146 by default, then goes down to 82 after I add the constraint.

**Comment by lwouis** [comment]
A bit unrelated but I thought I should share with you:

* https://www.logcg.com/en/archives/2902.html
* https://www.codesd.com/item/how-to-hang-up-remap-an-arbitrary-keyboard-event-on-osx.html

There is massive information in these articles. It made me curious to try alternate ways to record keys, such as [this project which uses HID API](https://github.com/SkrewEverything/Swift-Keylogger). I like their reasoning behind making it:

> Another problem is with Apple high-level APIs. I don't know the reason but Apple suddenly deprecates and removes the documentation of APIs from its website.
>
> So, I don't want to keep checking about the availability of their APIs and changing my code frequently. That is the reason I went for low-level API which is using HID API.

**Comment by Kentzo** [comment]
I have pushed an update that modifies the behavior of Global monitors with respect to the `isEnabled` property. As of now, handlers are added and removed dynamically depending on whether corresponding actions are enabled.

You may use this to dynamically opt in and opt out from global event handling.

**Comment by lwouis** [comment]
I think last time I tried `GlobalShortcutMonitor`, it didn't recognize modifiers-only `.keyUp`. For example `alt-up` to focus the window in AltTab wasn't recognized.

By the way, I released the new version of AltTab which uses ShortcutRecorder! The shortcut side seems great. There is still the issue with the size of the shortcut controls (the width=82px discussion), but I figured it's good enough to ship and give the users the new features :)

**Comment by Kentzo** [comment]
Correct, GlobalShortcutMonitor won't work. However I still hope you can intercept key events via a local events monitor.

Could you explain in detail why exactly you cannot capture keyboard focus? Perhaps I can come up with a sensible workaround.

**Comment by lwouis** [comment]
> Could you explain in detail why exactly you cannot capture keyboard focus?

I'll call the window the user was interacting with "the background window".

Having AltTab window become key means:

* The background window locus clues of visual focus: get's dimmed, color circles on the top left change. This can be distracting in some cases
* keyboard focus is lost. Let's say the user is typing text and it's cursor is in a specific spot he wants to paste into. He hits alt-tab to check another app but then he remembers and cancels alt-tab. At this point we can't bring the lost keyboard focus back to where it was. I think?

I checked other apps like HyperSwitch and they don't steal keyboard focus. Especially interesting is WindowSwitcher which doesn't lose focus *and* works on password fields, which is something that CGEvent taps don't allow, but that the Carbon APIs allow. I know because i tested with GlobalShortcutMonitor and it worked. And somehow WindowSwitcher has the modifier-only .keyUp work too. I wish i knew their implementation

**Comment by Kentzo** [comment]
> A bit unrelated but I just found this out: the allowsEmptyModifierFlags is not respected.

Should be fixed.

> That was very simple to do by overriding the following method in a subclass of RecorderControl:

I suggest to implement delegate's `recorderControl(_:, canRecord:)` instead.

**Comment by lwouis** [comment]
> However I still hope you can intercept key events via a local events monitor

Actually, why do you hope for that? The current solution i used is CGEvent tap for everything, then calling the localshortcutmonitor.handle method. Is there any benefit to using activating the app, then catching local events instead? Also there will always be a need for global event listening for initial activation.

What would be the benefit of switching away from cgevent taps? For the password issue, i'm not even sure it's desirable. I'm starting to think it's a security breach on Apple side and that i shouldn't have the app listen on those password fields even if i could

**Comment by Kentzo** [comment]
If AltTab depend on the Accessibility permission, then indeed there is no gain.

Otherwise my hope is to being able to help you to get rid of this permission. It is akin to asking for root: you can inject any event. Which makes your app a target for malware.

**Comment by lwouis** [comment]
AltTab is already making extensive use of the Accessibility APIs to see OS events such as "a new window was opened", "an app was quit", "the title of a window changed", or to focus windows. I don't think there is anyway we can stop granting these permissions. It makes sense in many ways that AltTab is an accessibility app.

Regarding being a target for malware, It's about what is done in the end. If i use a private or deprecated API that Apple is not restricting to do the same thing without asking the user for permission, one could argue that it's making the situation even worse since now we are keylogging the user for example, and they were not even shown the permission prompt.

I think the current cgevent tap implem is satisfying. I'm a bit worried about the control widths potentially confusing some users during interaction

**Comment by Kentzo** [comment]
I will be looking onto rendering today. In general I also suggest implementing custom style if custom constraints produce bad results.

**Comment by lwouis** [comment]
> I also suggest implementing custom style

I looked into the wiki section for styling, and also the playgrounds, but couldn't figure out how it works

**Comment by Kentzo** [comment]
Right now the best way is to read the SRRecorderControlStyle.h header.

**Comment by Kentzo** [comment]
The bug with label clipping should be addressed now.

**Comment by lwouis** [comment]
> The bug with label clipping should be addressed now.

I think there are a few issues. Here is the behavior I observe:

Control with `ctrl`:

![image](https://user-images.githubusercontent.com/106195/78500939-5d282a80-7794-11ea-80c0-8fcd10af7456.png)

I click then press `ctrl` and `alt`; it doesn't show `alt`:

![Screen Shot 2020-04-05 at 23 23 08](https://user-images.githubusercontent.com/106195/78500968-86e15180-7794-11ea-8156-07ec58a1720e.jpg)

On release, it then shows ellipsis even though there is lots of horizontal space:

![image](https://user-images.githubusercontent.com/106195/78501067-125ae280-7795-11ea-8004-22a830c0f0de.png)

If instead of `ctrl` + `alt`, I press `shift` then `alt`, I see in order:

![Screen Shot 2020-04-05 at 23 26 35](https://user-images.githubusercontent.com/106195/78501034-f2c3ba00-7794-11ea-8343-0ef30b184cbf.jpg)

![Screen Shot 2020-04-05 at 23 26 44](https://user-images.githubusercontent.com/106195/78501035-f3f4e700-7794-11ea-961d-8efa79492e3a.jpg)

Then on release:

![image](https://user-images.githubusercontent.com/106195/78501048-02db9980-7795-11ea-9de7-5634bb2ceb92.png)

Another case with lots of modifiers:

![image](https://user-images.githubusercontent.com/106195/78501104-361e2880-7795-11ea-9dd0-dfac32178e6c.png)

The ellipsis-after-release issue seems to only happen with `ctrl` + `alt` and `ctrl` + `cmd`. All other combinations don't show the issue, even `ctrl` + `alt` + `cmd`.

So to recap, there is an issue during recording, where the keys in-progress are not showing properly, then there is an issue after recording, where some combination show an ellipsis even if there is enough room.

**Comment by Kentzo** [comment]
> During recording, where the keys in-progress are not showing properly
> After recording, where some combination show an ellipsis even if there is enough room.
> With some keycode-only shortcuts with almost any keycode except tab

Please take a look now. Enabling label "shifting" exposed a couple of issues related to redrawing.

> The tooltip doesn't show with allowsExpansionToolTips = true

An "extension tooltip" is not the same as a tooltip: it draws on top of control's content, extending beyond its frame. The recorder control does not support this option out of the box as it does not use `NSCell`.

I will look into implementing this, but no promises.

**Comment by lwouis** [comment]
> Please take a look now. Enabling label "shifting" exposed a couple of issues related to redrawing.

I just tested the new code, and it seems to fix all issues. Fantastic work, thanks!

> An "extension tooltip" is not the same as a tooltip: it draws on top of control's content, extending beyond its frame. The recorder control does not support this option out of the box as it does not use NSCell.

Right now I have set the `widthAnchor` to `100`, and it seems to allow un-ellipsed displays of most shortcuts. I'm sure there is a way to make a larger shortcut that ellipses, but most people will experience a non-ellipsed UI I think.

Do you know if there is a way to have the control resize depending on content. Some kind of use of `intrinsicSize` perhaps?

**Comment by Kentzo** [comment]
The builtin style is designed to retain control's size fixed and does have intrinsic size that is only as big as needed to display the longest possible label.

**Comment by lwouis** [comment]
I meant, do you know of a way I could have the control follow the size of its *actual* label, not the longest possible label?

**Comment by Kentzo** [comment]
I'd suggest to experiment by deactivating [the suggested width constraint](https://github.com/Kentzo/ShortcutRecorder/blob/c0b3a415c887a1b49c2cd4c497bfa8543eb60132/Library/SRRecorderControlStyle.m#L1076-L1081). The control then should resize itself with respect to the label width.

This is an internal detail, so no promises that I will be retain for the releases to come.

**Comment by lwouis** [comment]
I tried disabling this constraint, but it didn't have any visible effect. I then tried disabling all constraints in my `ShortcutRecorder` subclass with:

```swift
convenience init(_ shortcutString: String, _ modifierFlagsOnly: Bool) {
    self.init(frame: .zero)
    [...]
    style.alwaysConstraints.forEach { $0.isActive = false }
    style.displayingConstraints.forEach { $0.isActive = false }
}
```

Even this doesn't do anything. Maybe it's enabled later and I'm too early or at the wrong time in the lifecycle of the class?

**Comment by Kentzo** [comment]
You have to subclass style to disable the constraint and exclude if from the `*constraints` properties.

**Comment by lwouis** [comment]
I subclassed `RecorderControlStyle` and disable the constraints, but still not results. Also I can remove the constraint from the `alwaysConstraints` array because it is `readonly`. I tried to look for a way to do this in the `RecorderControlStyle.m` file but it seems that there is an internal `_ alwaysConstraints` variable that never exposed, that's later synthesized as a readonly array `alwaysConstraints`.

```swift
class FitWidthStyle: RecorderControlStyle {
    convenience init(_ identifier: String?, _ components: RecorderControlStyle.Components?) {
        self.init(identifier: identifier, components: components)
        // alwaysConstraints.first(where: { $0.identifier == "SR_alignmentGuide_suggestedWidth"})!.isActive = false
        alwaysConstraints.forEach { $0.isActive = false }
        displayingConstraints.forEach { $0.isActive = false }
    }
}
```

**Comment by Kentzo** [comment]
`init` is too soon. You need to deactivate the constraint in `addConstraints` then modify the getter of `alwaysConstraints` to not return it (otherwise it will be re-activated by the control).

**Comment by lwouis** [comment]
> init is too soon. You need to deactivate the constraint in addConstraints then modify the getter of alwaysConstraints to not return it (otherwise it will be re-activated by the control).

I tried to implement something around those guidelines. I couldn't get it to work. I think it's fine for now, especially since it would be messing with internals and making it risky on updates later down the line.

I'm actually surprised nobody asked for a way to have a smaller control before. The current one is quite large by default

By the way, regarding this ticket, I think you could close it. I have released a version of AltTab with ShortcutRecorder finally integrated a few days ago, and it seems to be working great for users 👍

**Comment by Kentzo** [comment]
I have removed the explicit constraint for intrinsic width. The control now relies on the `intrinsicContentSize` provided by the style alone. In other words it should be enough to override `intrinsicContentSize` and return `NSViewNoIntrinsicMetric` for width.

Constraint for the height is still present as it's defined by the background assets.

Other things you may try to make the control smaller:

- Custom style with other font / smaller font size
- Adjust margins in style metrics
- Smaller assets for background
- Return a shorter string in `noValueNormalLabel`, `noValueRecordingLabel` and `noValueDisableLabel`

**Comment by Kentzo** [comment]
Feel free to create a new issue to include a smaller style into the framework. I did that in the past and it worked. I included a complimentary "bonus" of $150 though :)

**Comment by Kentzo** [comment]
@lwouis Can you check if `SRAXGlobalShortcutMonitor` works for you when you change `kCGEventTapOptionDefault` to `kCGEventTapOptionListenOnly`?

**Comment by lwouis** [comment]
@Kentzo I'm not sure what you mean. Currently in AltTab, [I use `kCGEventTapOptionDefault`](https://github.com/lwouis/alt-tab-macos/blob/master/src/logic/events/KeyboardEvents.swift#L19), which i need since I do modify events (i.e. absorb them if they match within ShortcutRecorder).

**Comment by Kentzo** [comment]
Ah I see, indeed you do need `kCGEventTapOptionDefault`.

I'm considering what'd be a better default for ShortcutRecorder in general.

**Comment by lwouis** [comment]
I would say least-priviledge is nice as a default, as long as people can opt-in the `default` mode, it's fine to have `listenonly` be the default, no?

**Comment by Kentzo** [comment]
My thoughts exactly. Please take a look at the develop branch.

**Comment by lwouis** [comment]
You mean that commit, right?

https://github.com/Kentzo/ShortcutRecorder/commit/ae35edde1018130662443b80aa3799c28bccdc48

I checked it and it seems nice to me. You let users set the `RunLoop` and also active/passive event filtering. Seems all good to me.

You may want to note that some users in AltTab complained that [sometimes the keyboard shortcuts stop working](https://github.com/lwouis/alt-tab-macos/issues/157#issuecomment-627162866). I have tried my best but so far I have no idea how this is happening. It may simply be a macOS bug in that `CGEvent` api. So keep that in mind perhaps.

**Comment by Kentzo** [comment]
> I still don't understand to this day for example, why these are happening; what triggers them.

You may get that type, even if you didn't register for it explicitly, e.g. if system decides your handler takes too long. Installed "tap" appears to be called synchronously by macOS''s input system.

Thus I suggest to avoid doing any work in the handler and instead handle asynchronously.

> You may want to note that some users in AltTab complained that sometimes the keyboard shortcuts stop working.

Perhaps that's exactly what was happening?

**Comment by lwouis** [comment]
> if system decides your handler takes too long to execute.

Do you have any proof that this actually happens? It would seem like such an oversight from Apple not to document it if that's the case. Very interesting in any case. I'll try to make the code returns as fast as possible. Thank you for the info!

**Comment by Kentzo** [comment]
I have encountered it myself while working on this feature. I then added the necessary ‘if’.

The doc in the header says that these messages are out of band.

**Comment by lwouis** [comment]
Hi @Kentzo!

So, we've been trying to identify a [keyboard issue](https://github.com/lwouis/alt-tab-macos/issues/157) with AltTab's community for a while now. Today, I analyzed new logs from a user, and I see this message in these logs:

```
#Error Unexpected key code 0 for the FlagsChanged event
```

This message is coming from ShortcutRecorder. After digging a bit, I found the log line:

https://github.com/Kentzo/ShortcutRecorder/blob/efde820f44562e4073924142694c0b6923c30597/Library/SRShortcutAction.m#L395

I'm quite puzzled as to how this is happening. From the caller (AltTab), we call get a `CGEvent` from `CGEvent.tapCreate`. Then we construct a `NSEvent` based on that `CGEvent`:

```swift
let event_ = NSEvent(cgEvent: cgEvent)
```

Then we create a new `NSEvent` based off that `NSEvent` (to workaround the issue we had with calling `.characters` outside the main thread; I think you remember this workaround):

```swift
let event = NSEvent.keyEvent(with: event_.type, location: event_.locationInWindow, modifierFlags: event_.modifierFlags, timestamp: event_.timestamp, windowNumber: event_.windowNumber, context: nil, characters: "", charactersIgnoringModifiers: "", isARepeat: type == .flagsChanged ? false : event_.isARepeat, keyCode: event_.keyCode)
```

At this point we call `event.sr_keyEventType` to learn if a modifier key was went up or down. Within this method, the log line is logged.

I can't imagine why the `keyCode` would be `0`. From the NSEvent header, it doesn't even seem like 0 should be a possible value:

```
/* Device-independent bits found in event modifier flags */
typedef NS_OPTIONS(NSUInteger, NSEventModifierFlags) {
    NSEventModifierFlagCapsLock           = 1 << 16, // Set if Caps Lock key is pressed.
    NSEventModifierFlagShift              = 1 << 17, // Set if Shift key is pressed.
    NSEventModifierFlagControl            = 1 << 18, // Set if Control key is pressed.
    NSEventModifierFlagOption             = 1 << 19, // Set if Option or Alternate key is pressed.
    NSEventModifierFlagCommand            = 1 << 20, // Set if Command key is pressed.
    NSEventModifierFlagNumericPad         = 1 << 21, // Set if any key in the numeric keypad is pressed.
    NSEventModifierFlagHelp               = 1 << 22, // Set if the Help key is pressed.
    NSEventModifierFlagFunction           = 1 << 23, // Set if any function key is pressed.
    
    // Used to retrieve only the device-independent modifier flags, allowing applications to mask off the device-dependent modifier flags, including event coalescing information.
    NSEventModifierFlagDeviceIndependentFlagsMask    = 0xffff0000UL
};
```

Do you have any idea how this scenario would be happening?

**Comment by Kentzo** [comment]
Can you catch that case before conversion to NSEvent and dump the CGEvent (e.g. via CGEventCreateData)?

**Comment by Kentzo** [comment]
@lwouis Btw, try https://github.com/Kentzo/ShortcutRecorder/blob/develop/Library/SRShortcut.h#L80 to avoid some ugliness in your code to circumvent that `characters` bug.
